xs/src/libslic3r/ExPolygon.cpp

```diff
@@ -195,107 +291,258 @@ ExPolygon::medial_axis(const ExPolygon &bounds, double max_width, double min_wid
     ThickPolylines pp;
     ma.build(&pp);
     
-    /* // Commented out debug code
-    SVG svg("medial_axis.svg");
-    svg.draw(*this);
-    svg.draw(pp);
-    svg.Close();
-    */
+
+    //{
+    //    stringstream stri;
+    //    stri << "medial_axis" << id << ".svg";
+    //    SVG svg(stri.str());
+    //    svg.draw(bounds);
+    //    svg.draw(*this);
+    //    svg.draw(pp);
+    //    svg.Close();
+    //}
     
     // Find the maximum width returned; we're going to use this for validating and 
     // filtering the output segments.
     double max_w = 0;
     for (ThickPolylines::const_iterator it = pp.begin(); it != pp.end(); ++it)
         max_w = fmaxf(max_w, *std::max_element(it->width.begin(), it->width.end()));
-    
-    
-    //  Aligned fusion: Fusion the bits at the end of lines by "increasing thickness"
-    //  For that, we have to find other lines,
-    //  and with a next point no more distant than the max width.
-    //  Then, we can merge the bit from the first point to the second by following the mean.
 
+    concatThickPolylines(pp);
+    //reoder pp by length (ascending) It's really important to do that to avoid building the line from the width insteand of the length
+    std::sort(pp.begin(), pp.end(), [](const ThickPolyline & a, const ThickPolyline & b) { return a.length() < b.length(); });
+
+    // Aligned fusion: Fusion the bits at the end of lines by "increasing thickness"
+    // For that, we have to find other lines,
+    // and with a next point no more distant than the max width.
+    // Then, we can merge the bit from the first point to the second by following the mean.
+    //
+    int id_f = 0;
     bool changes = true;
     while (changes) {
         changes = false;
         for (size_t i = 0; i < pp.size(); ++i) {
             ThickPolyline& polyline = pp[i];
+            
+            //simple check to see if i can be fusionned
+            if (!polyline.endpoints.first && !polyline.endpoints.second) continue;
 
             ThickPolyline* best_candidate = nullptr;
             float best_dot = -1;
             int best_idx = 0;
-
+            double dot_poly_branch = 0;
+            double dot_candidate_branch = 0;
+            
             // find another polyline starting here
             for (size_t j = i + 1; j < pp.size(); ++j) {
```

Comment:
Maybe it would be worth finding all polylines starting from (ends of) `polyline`, assign scores to them and select good candidates ... It would partially remove `k` loop and may be easier to understand the algorithm ...